consensus/hotstuff/safetyrules/safety_rules_test.go

```diff
@@ -447,29 +483,75 @@ func (s *SafetyRulesTestSuite) TestProduceTimeout_ShouldTimeout() {
 // Hence, the cases tested here would be symptoms of an internal bugs, and therefore should not result in an NoVoteError.
 func (s *SafetyRulesTestSuite) TestProduceTimeout_NotSafeToTimeout() {
 
-	s.Run("highest-qc-below-locked-round", func() {
-		view := s.proposal.Block.View
-		newestQC := helper.MakeQC(helper.WithQCView(s.safetyData.LockedOneChainView - 1))
+	s.Run("newest-qc-nil", func() {
+		// newestQC cannot be nil
+		timeout, err := s.safety.ProduceTimeout(s.safetyData.LockedOneChainView, nil, nil)
+		require.Error(s.T(), err)
+		require.False(s.T(), model.IsNoVoteError(err))
+		require.Nil(s.T(), timeout)
+	})
+	s.Run("happy-path-includes-last-view-tc", func() {
+		newestQC := helper.MakeQC(helper.WithQCView(s.safetyData.LockedOneChainView))
 
-		timeout, err := s.safety.ProduceTimeout(view, newestQC, nil)
+		// tc not needed but included
+		timeout, err := s.safety.ProduceTimeout(newestQC.View+1, newestQC, helper.MakeTC())
 		require.Error(s.T(), err)
 		require.False(s.T(), model.IsNoVoteError(err))
 		require.Nil(s.T(), timeout)
 	})
-	s.Run("cur-view-below-highest-acknowledged-view", func() {
-		view := s.safetyData.HighestAcknowledgedView
+	s.Run("last-view-tc-nil", func() {
+		newestQC := helper.MakeQC(helper.WithQCView(s.safetyData.LockedOneChainView))
+
+		// tc needed but not included
+		timeout, err := s.safety.ProduceTimeout(newestQC.View+2, newestQC, nil)
+		require.Error(s.T(), err)
+		require.False(s.T(), model.IsNoVoteError(err))
+		require.Nil(s.T(), timeout)
+	})
+	s.Run("last-view-tc-for-wrong-view", func() {
 		newestQC := helper.MakeQC(helper.WithQCView(s.safetyData.LockedOneChainView))
+		// lastViewTC should be for newestQC.View+1
+		lastViewTC := helper.MakeTC(helper.WithTCView(newestQC.View))
 
-		timeout, err := s.safety.ProduceTimeout(view-2, newestQC, nil)
+		timeout, err := s.safety.ProduceTimeout(newestQC.View+2, newestQC, lastViewTC)
 		require.Error(s.T(), err)
 		require.False(s.T(), model.IsNoVoteError(err))
 		require.Nil(s.T(), timeout)
 	})
 	s.Run("cur-view-below-highest-QC", func() {
```

Comment:
The boundary condition is `curView==highestQC.view`, we could test that in addition, or instead.

Reply:
This test checks curView > highestQC.View, I will add extra test for curView == highestQC.View